### ISSUE TITLE
Tactile setup

### ISSUE THREAD

**Issue by TheSuperpapi** [pibooth/pibooth#188]
Hello there,

Here is what was updated to ptopose this tactile interface.

[Click on me](https://youtu.be/FaiWML-OOus)

In the background.py (just replace the .png file by what you want):
```
# -*- coding: utf-8 -*-

import os.path as osp
import pygame

from pibooth import fonts, pictures
from pibooth.language import get_translated_text

ARROW_TOP = 'top'
ARROW_BOTTOM = 'bottom'
ARROW_HIDDEN = 'hidden'


def multiline_text_to_surfaces(text, color, rect, align='center'):
    """Return a list of surfaces corresponding to each line of the text.
    The surfaces are next to each others in order to fit the given rect.

    The ``align`` parameter can be one of:
       * top-left
       * top-center
       * top-right
       * center-left
       * center
       * center-right
       * bottom-left
       * bottom-center
       * bottom-right
    """
    surfaces = []
    lines = text.splitlines()

    font = fonts.get_pygame_font(max(lines, key=len), fonts.CURRENT,
                                 rect.width, rect.height / len(lines))
    for i, line in enumerate(lines):
        surface = font.render(line, True, color)

        if align.endswith('left'):
            x = rect.left
        elif align.endswith('center'):
            x = rect.centerx - surface.get_rect().width / 2
        elif align.endswith('right'):
            x = rect.right - surface.get_rect().width / 2
        else:
            raise ValueError("Invalid horizontal alignment '{}'".format(align))

        height = surface.get_rect().height
        if align.startswith('top'):
            y = rect.top + i * height
        elif align.startswith('center'):
            y = rect.centery - len(lines) * height / 2 + i * height
        elif align.startswith('bottom'):
            y = rect.bottom - (len(lines) - i) * height
        else:
            raise ValueError("Invalid vertical alignment '{}'".format(align))

        surfaces.append((surface, surface.get_rect(x=x, y=y)))
    return surfaces


class Background(object):

    def __init__(self, image_name, color=(0, 0, 0), text_color=(255, 255, 255)):
        self._rect = None
        self._outlines = True
        self._name = image_name
        self._need_update = False

        self._background = None
        self._background_color = color
        self._background_image = None

        self._overlay = None

        self._texts = []  # List of (surface, rect)
        self._text_border = 20  # Distance to other elements
        self._text_color = text_color

    def __str__(self):
        """Return background final name.

        It is used in the main window to distinguish backgrounds in the cache
        thus each background string shall be uniq.
        """
        return "{}({})".format(self.__class__.__name__, self._name)

    def _write_text(self, text, rect=None, align='center'):
        """Write a text in the given rectangle.
        """
        if not rect:
            rect = self._rect.inflate(-self._text_border, -self._text_border)
        if self._outlines:
            # Build rectangle around text area for debuging purpose
            outlines = pygame.Surface(rect.size, pygame.SRCALPHA, 32)
            pygame.draw.rect(outlines, pygame.Color(255, 0, 0), outlines.get_rect(), 2)
            self._texts.append((outlines, rect))
        self._texts.extend(multiline_text_to_surfaces(text, self._text_color, rect, align))

    def set_color(self, color_or_path):
        """Set background color (RGB tuple) or path to an image that used to
        fill the background.

        :param color_or_path: RGB color tuple or image path
        :type color_or_path: tuple or str
        """
        if isinstance(color_or_path, (tuple, list)):
            assert len(color_or_path) == 3, "Length of 3 is required for RGB tuple"
            if color_or_path != self._background_color:
                self._background_color = color_or_path
                self._need_update = True
        else:
            assert osp.isfile(color_or_path), "Invalid image for window background: '{}'".format(color_or_path)
            if color_or_path != self._background_image:
                self._background_image = color_or_path
                self._need_update = True

    def set_text_color(self, color):
        """Set text color (RGB tuple) used to write the texts.

        :param color: RGB color tuple
        :type color: tuple
        """
        assert len(color) == 3, "Length of 3 is required for RGB tuple"
        if color != self._text_color:
            self._text_color = color
            self._need_update = True

    def set_outlines(self, outlines=True):
        """Draw outlines for each rectangle available for drawing
        texts.

        :param outlines: enable / disable outlines
        :type outlines: bool
        """
        if outlines != self._outlines:
            self._outlines = outlines
            self._need_update = True

    def resize(self, screen):
        """Resize objects to fit to the screen.
        """
        if self._rect != screen.get_rect():
            self._rect = screen.get_rect()

            if self._background_image:
                self._background = pictures.get_pygame_image(
                    self._background_image, (self._rect.width, self._rect.height), crop=True, color=None)
                self._background_color = pictures.get_pygame_main_color(self._background)

            overlay_name = "{}.png".format(self._name)
            if osp.isfile(pictures.get_filename(overlay_name)):
                self._overlay = pictures.get_pygame_image(
                    pictures.get_filename(overlay_name), (self._rect.width, self._rect.height), color=self._text_color, bg_color=self._background_color)


            self.resize_texts()

            self._need_update = True

    def resize_texts(self, rect=None, align='center'):
        """Update text surfaces.
        """
        self._texts = []
        text = get_translated_text(self._name)
        if text:
            self._write_text(text, rect, align)

    def paint(self, screen):
        """Paint and animate the surfaces on the screen.
        """
        if self._background:
            screen.blit(self._background, (0, 0))
        else:
            screen.fill(self._background_color)
        if self._overlay:
            screen.blit(self._overlay, self._overlay.get_rect(center=self._rect.center))
        for text_surface, pos in self._texts:
            screen.blit(text_surface, pos)
        self._need_update = False


class IntroBackground(Background):

    def __init__(self, arrow_location=ARROW_BOTTOM, arrow_offset=0):
        Background.__init__(self, "intro")
        self.arrow_location = arrow_location
        self.arrow_offset = arrow_offset
        self.left_arrow = None
        self.left_arrow_pos = None

    def resize(self, screen):
        Background.resize(self, screen)
        if self._need_update and self.arrow_location != ARROW_HIDDEN:
            size = (self._rect.width * 0.5, self._rect.height * 0.5)

            vflip = True if self.arrow_location == ARROW_TOP else False
            self.left_arrow = pictures.get_pygame_image("picture.png", size, vflip=vflip, color=self._text_color)

            x = int(self._rect.left + self._rect.width // 4
                    - self.left_arrow.get_rect().width // 2)
            if self.arrow_location == ARROW_TOP:
                y = self._rect.top + 10
            else:
                y = int(self._rect.top + 1 * self._rect.height // 3)

            self.left_arrow_pos = (x * 1.3 - self.arrow_offset, y)

    def resize_texts(self):
        """Update text surfaces.
        """
        if self.arrow_location == ARROW_HIDDEN:
            rect = pygame.Rect(self._text_border, self._text_border,
                               self._rect.width / 2 - 2 * self._text_border,
                               self._rect.height - 2 * self._text_border)
            align = 'center'
        elif self.arrow_location == ARROW_BOTTOM:
            rect = pygame.Rect(self._text_border, self._text_border,
                               self._rect.width / 2 - 2 * self._text_border,
                               self._rect.height * 0.4 - self._text_border)
            align = 'bottom-center'
        else:
            rect = pygame.Rect(self._text_border, self._rect.height * 0.4,
                               self._rect.width / 2 - 2 * self._text_border,
                               self._rect.height * 0.6 - self._text_border)
            align = 'top-center'
        Background.resize_texts(self, rect, align)

    def paint(self, screen):
        Background.paint(self, screen)
        if self.arrow_location != ARROW_HIDDEN:
            screen.blit(self.left_arrow, self.left_arrow_pos)


class IntroWithPrintBackground(IntroBackground):

    def __init__(self, arrow_location=ARROW_BOTTOM, arrow_offset=0):
        IntroBackground.__init__(self, arrow_location, arrow_offset)
        self.right_arrow = None
        self.right_arrow_pos = None

    def __str__(self):
        """Return background final name.

        It is used in the main window to distinguish backgrounds in the cache
        thus each background string shall be uniq.
pi        """
        return "{}({})".format(self.__class__.__name__, "intro_print")

    def resize(self, screen):
        IntroBackground.resize(self, screen)
        if self._need_update and self.arrow_location != ARROW_HIDDEN:
            size = (self._rect.width * 0.07, self._rect.height * 0.07)
            vflip = True if self.arrow_location == ARROW_TOP else False
            angle = 0 if self.arrow_location == ARROW_TOP else 0
            self.right_arrow = pictures.get_pygame_image("printer.png", size, hflip=False,
                                                         vflip=vflip, angle=angle, color=self._text_color)
            x = int(self._rect.left + self._rect.width // 2
                    - self.right_arrow.get_rect().width // 2)
            if self.arrow_location == ARROW_TOP:
                y = self._rect.top + 10
            else:
                y = int(self._rect.bottom - self.right_arrow.get_rect().height * 1.2)
            self.right_arrow_pos = (x * 1.995 - self.arrow_offset, y)

    def resize_texts(self):
        """Update text surfaces.
        """
        IntroBackground.resize_texts(self)
        text = get_translated_text("intro_print")
        if text:
            if self.arrow_location == ARROW_HIDDEN or self.arrow_location == ARROW_BOTTOM:
                rect = pygame.Rect(self._rect.width * 0.30 + self._text_border, 0,
                                   self._rect.width * 0.20 - 2 * self._text_border,
                                   self._rect.height * 0.3 - 2 * self._text_border)
                rect.bottom = self._rect.height - self._rect.height * 0.08
            else:
                rect = pygame.Rect(self._rect.width * 0.30 + self._text_border, 0,
                                   self._rect.width * 0.20 - 2 * self._text_border,
                                   self._rect.height * 0.3 - 2 * self._text_border)
                rect.top = self._rect.height * 0.08
            self._write_text(text, rect)

    def paint(self, screen):
        IntroBackground.paint(self, screen)
        if self.arrow_location != ARROW_HIDDEN:
            screen.blit(self.right_arrow, self.right_arrow_pos)


class ChooseBackground(Background):

    def __init__(self, choices, arrow_location=ARROW_BOTTOM, arrow_offset=0):
        Background.__init__(self, "choose")
        self.arrow_location = arrow_location
        self.arrow_offset = arrow_offset
        self.choices = choices
        self.layout0 = None
        self.layout0_pos = None
        self.layout1 = None
        self.layout1_pos = None
        self.left_arrow = None
        self.left_arrow_pos = None
        self.right_arrow = None
        self.right_arrow_pos = None

    def resize(self, screen):
        Background.resize(self, screen)
        if self._need_update:
            size = (self._rect.width * 0.6, self._rect.height * 0.6)
            self.layout0 = pictures.get_pygame_layout_image(
                self._text_color, self._background_color, self.choices[0], size)
            self.layout1 = pictures.get_pygame_layout_image(
                self._text_color, self._background_color, self.choices[1], size)

            inter = (self._rect.width - 2 * self.layout0.get_rect().width) // 3

            x0 = int(self._rect.left + inter)
            x1 = int(self._rect.left + 2 * inter + self.layout0.get_rect().width)
            y = int(self._rect.top + self._rect.height * 0.3)

            self.layout0_pos = (x0, y)
            self.layout1_pos = (x1, y)

            if self.arrow_location != ARROW_HIDDEN:
                if self.arrow_location == ARROW_TOP:
                    y = 5
                    x_offset = 30
                    size = (self._rect.width * 0.1, self._rect.top + y + 30)
                else:
                    x_offset = 0
                    y = self.layout0_pos[1] + self.layout0.get_rect().height + 5
                    size = (self._rect.width * 0.1, self._rect.bottom - y - 5)

                vflip = True if self.arrow_location == ARROW_TOP else False
                self.left_arrow = pictures.get_pygame_image("empty.png", size, vflip=vflip,
                                                            color=self._text_color)
                self.right_arrow = pictures.get_pygame_image("empty.png", size, hflip=True,
                                                             vflip=vflip, color=self._text_color)

                inter = (self._rect.width - 2 * self.left_arrow.get_rect().width) // 4

                x0 = int(self._rect.left + inter) - x_offset
                x1 = int(self._rect.left + 3 * inter + self.left_arrow.get_rect().width) + x_offset

                self.left_arrow_pos = (x0 - self.arrow_offset, y)
                self.right_arrow_pos = (x1 + self.arrow_offset, y)

    def resize_texts(self):
        """Update text surfaces.
        """
        rect = pygame.Rect(self._text_border, self._text_border,
                           self._rect.width - 2 * self._text_border, self._rect.height * 0.2)
        Background.resize_texts(self, rect)

    def paint(self, screen):
        Background.paint(self, screen)
        screen.blit(self.layout0, self.layout0_pos)
        screen.blit(self.layout1, self.layout1_pos)
        if self.arrow_location != ARROW_HIDDEN:
            screen.blit(self.left_arrow, self.left_arrow_pos)
            screen.blit(self.right_arrow, self.right_arrow_pos)


class ChosenBackground(Background):

    def __init__(self, choices, selected):
        Background.__init__(self, "chosen")
        self.choices = choices
        self.selected = selected
        self.layout = None
        self.layout_pos = None

    def __str__(self):
        """Return background final name.
        It is used in the main window to distinguish background in the cache.
        """
        return "{}({}{})".format(self.__class__.__name__, self._name, self.selected)

    def resize(self, screen):
        Background.resize(self, screen)
        if self._need_update:
            size = (self._rect.width * 0.6, self._rect.height * 0.6)

            self.layout = pictures.get_pygame_layout_image(
                self._text_color, self._background_color, self.selected, size)

            x = self.layout.get_rect(center=self._rect.center).left
            y = int(self._rect.top + self._rect.height * 0.3)

            self.layout_pos = (x, y)

    def resize_texts(self):
        """Update text surfaces.
        """
        rect = pygame.Rect(self._text_border, self._text_border,
                           self._rect.width - 2 * self._text_border, self._rect.height * 0.2)
        Background.resize_texts(self, rect)

    def paint(self, screen):
        Background.paint(self, screen)
        screen.blit(self.layout, self.layout_pos)


class CaptureBackground(Background):

    def __init__(self):
        Background.__init__(self, "capture")
        self.left_people = None
        self.left_people_pos = None
        self.right_people = None
        self.right_people_pos = None

    def resize(self, screen):
        Background.resize(self, screen)
        if self._need_update:
            images_height = self._rect.height / 4
            size = (3 * images_height, images_height)

            self.left_people = pictures.get_pygame_image("capture_left_image.png", size=size,
                                                         color=self._text_color)
            self.right_people = pictures.get_pygame_image("capture_right_image.png", size=size,
                                                          color=self._text_color)

            x = int(self._rect.right - 2 * self.right_people.get_rect().width)
            y = int(self._rect.bottom - images_height)

            self.left_people_pos = (0, y)
            self.right_people_pos = (x, y)

    def paint(self, screen):
        Background.paint(self, screen)
        screen.blit(self.left_people, self.left_people_pos)
        screen.blit(self.right_people, self.right_people_pos)



class ProcessingBackground(Background):

    def __init__(self):
        Background.__init__(self, "processing")

    def resize_texts(self):
        """Update text surfaces.
        """
        rect = pygame.Rect(self._text_border, self._rect.height * 0.8 - self._text_border,
                           self._rect.width - 2 * self._text_border, self._rect.height * 0.2)
        Background.resize_texts(self, rect)


class PrintBackground(Background):

    def __init__(self, arrow_location=ARROW_BOTTOM, arrow_offset=0):
        Background.__init__(self, "print")
        self.arrow_location = arrow_location
        self.arrow_offset = arrow_offset
        self.right_arrow = None
        self.right_arrow_pos = None
        self.left_arrow = None
        self.left_arrow_pos = None

    def resize(self, screen):
        Background.resize(self, screen)
        if self._need_update and self.arrow_location != ARROW_HIDDEN:
            size = (self._rect.width * 0.5, self._rect.height * 0.5)

            vflip = True if self.arrow_location == ARROW_TOP else False

            # Right arrow
            self.right_arrow = pictures.get_pygame_image(
                "printer2.png", size, hflip=True, vflip=vflip, color=self._text_color)

            x = int(self._rect.left + self._rect.width * 0.75
                    - self.right_arrow.get_rect().width // 2)

            if self.arrow_location == ARROW_TOP:
                y = self._rect.top + 10
            else:
                y = int(self._rect.top + 0.8 * self._rect.height // 3)

            self.right_arrow_pos = (x * 0.9 + self.arrow_offset, y)

            # Left arrow
            size = (self._rect.width * 0.2, self._rect.height * 0.2)
            angle = 70 if self.arrow_location == ARROW_TOP else -0
            self.left_arrow = pictures.get_pygame_image(
                "trash.png", size, hflip=False, vflip=vflip, angle=angle, color=self._text_color)

            x = int(self._rect.left + self._rect.width // 2
                    - self.left_arrow.get_rect().width // 2)

            if self.arrow_location == ARROW_TOP:
                y = self._rect.top + 10
            else:
                y = int(self._rect.bottom - self.left_arrow.get_rect().height * 1.1)

            self.left_arrow_pos = (7, y)

    def resize_texts(self):
        """Update text surfaces.
        """
        if self.arrow_location == ARROW_HIDDEN:
            rect = pygame.Rect(self._rect.width / 2 + self._text_border, self._text_border,
                               self._rect.width / 2 - 2 * self._text_border,
                               self._rect.height - 2 * self._text_border)
            align = 'center'
        elif self.arrow_location == ARROW_BOTTOM:
            rect = pygame.Rect(self._rect.width / 2 + self._text_border, self._text_border,
                               self._rect.width / 2 - 2 * self._text_border,
                               self._rect.height * 0.4 - self._text_border)
            align = 'bottom-center'
        else:
            rect = pygame.Rect(self._rect.width / 2 + self._text_border, self._rect.height * 0.4,
                               self._rect.width / 2 - 2 * self._text_border,
                               self._rect.height * 0.6 - self._text_border)
            align = 'top-center'
        Background.resize_texts(self, rect, align)

        text = get_translated_text("print_forget")
        if text:
            if self.arrow_location == ARROW_HIDDEN or self.arrow_location == ARROW_BOTTOM:
                rect = pygame.Rect(self._rect.width // 2, 0,
                                   self._rect.width // 5 - 2 * self._text_border,
                                   self._rect.height * 0.3 - 2 * self._text_border)
                rect.bottom = self._rect.height - self._rect.height * 0.08
            else:
                rect = pygame.Rect(self._rect.width // 2, 0,
                                   self._rect.width // 5 - 2 * self._text_border,
                                   self._rect.height * 0.3 - 2 * self._text_border)
                rect.top = self._rect.height * 0.08

            self._write_text(text, rect)

    def paint(self, screen):
        Background.paint(self, screen)
        if self.arrow_location != ARROW_HIDDEN:
            screen.blit(self.right_arrow, self.right_arrow_pos)
            screen.blit(self.left_arrow, self.left_arrow_pos)


class FinishedBackground(Background):

    def __init__(self):
        Background.__init__(self, "finished")

    def resize_texts(self):
        """Update text surfaces.
        """
        rect = pygame.Rect(self._rect.width * 0.3, self._rect.height * 0.6 - self._text_border,
                           self._rect.width * 0.4, self._rect.height * 0.4)
        Background.resize_texts(self, rect)


class OopsBackground(Background):

    def __init__(self):
        Background.__init__(self, "oops")

```

Globaly, as you can see, I only replaced the arrow PNG by other PNG and changed size and position.
If you need more, please ask.

Sorry for the delay and this post quality, but without internet I spent 1H to make and post it.... hope my internet issue will be solved soon.

Best Regards to all of you.
Papi

**Comment by werdeil** [comment]
Thanks a lot @TheSuperpapi we will see how to integrate your work on a Pull request.

**Comment by sravel** [comment]
yes great job @TheSuperpapi . I would also like to test because I find it nice to have the tactile side, is it possible to share your images from the video?

**Comment by TheSuperpapi** [comment]
Hello, here are the tactil .png icon,

Let me know if you can open and use this zip file.

You must copy and past it in the right folder.

[Tactil png icon.zip](https://github.com/pibooth/pibooth/files/5377167/Tactil.png.icon.zip)

See you

**Comment by werdeil** [comment]
I have pushed a PR to have a touchscreen option in #191. I've used other icon to be sure to credit the owners.

Here is an example of how it renders: https://youtu.be/PpSD7dVkaiM

**Comment by anxuae** [comment]
Hi @TheSuperpapi ,

The PR is now merged, the feature will be available in pibooth 2.0.2

Thanks for reporting.
anxuae
